src/v/cluster/self_test/netcheck.cc

```diff
@@ -0,0 +1,186 @@
+}
+
+ss::future<self_test_result>
+netcheck::run_individual_benchmark(model::node_id peer) {
+    auto irange = boost::irange<uint16_t>(0, _opts.parallelism);
+    static const auto two_seconds_us = 200000;
+    metrics m{two_seconds_us};
+    vassert(
+      _opts.max_duration >= _opts.duration, "Misconfigured self_test plan");
+    const auto max_deadline = ss::lowres_clock::now() + _opts.max_duration;
+    auto durations = co_await ssx::parallel_transform(
+      irange, [this, max_deadline, peer, &m](auto) {
+          return run_benchmark_fiber(
+            run_fiber_opts(max_deadline, _opts.duration), peer, m);
+      });
+
+    /// Total time will be the average time all fibers took to complete
+    const auto total_duration = std::accumulate(
+      durations.begin(), durations.end(), ss::lowres_clock::duration{0});
+    m.set_total_time(total_duration / _opts.parallelism);
+    auto result = m.to_st_result();
+    result.name = _opts.name;
+    result.info = fmt::format("Test performed against node: {}", peer);
+    if (_cancelled) {
+        result.warning = "Run was manually cancelled";
+    }
+    co_return result;
+}
+
+ss::future<ss::lowres_clock::duration> netcheck::run_benchmark_fiber(
+  run_fiber_opts fiber_state, model::node_id peer, metrics& m) {
+    /// run_fiber_opts manages the the RPC request timeouts to ensure that
+    /// the maximum amount of time for an individual RPC never exceeds the
+    /// remaining time of the test
+    while (!_cancelled && fiber_state.should_continue()) {
+        if (unlikely(_as.abort_requested())) {
+            throw netcheck_aborted_exception();
+        }
+        auto req = co_await make_netcheck_request(_self, _opts.request_size);
+        co_await m.measure(
+          [this, req = std::move(req), peer, &fiber_state, &m]() mutable {
+              return _connections.local()
+                .with_node_client<self_test_rpc_client_protocol>(
+                  _self,
+                  ss::this_shard_id(),
+                  peer,
+                  1s,
+                  [req = std::move(req)](
+                    self_test_rpc_client_protocol c) mutable {
+                      return c.netcheck(
+                        std::move(req), rpc::client_opts(200ms));
+                  })
+                .then([this, &fiber_state, &m](auto reply) {
+                    return process_netcheck_reply(
+                      std::move(reply), fiber_state, m);
+                });
+          });
+    }
+    co_return fiber_state.end - fiber_state.begin;
+}
+
+ss::future<size_t> netcheck::process_netcheck_reply(
+  result<rpc::client_context<netcheck_response>> reply,
+  run_fiber_opts& fiber_state,
+  metrics& m) {
+    if (!reply) {
+        if (
+          reply.error() == rpc::errc::client_request_timeout
+          || reply.error() == rpc::errc::exponential_backoff) {
+            co_await ss::sleep_abortable(200ms, _as);
+            throw omit_metrics_measurement_exception();
```

Comment:
This is just skipping the measurement, right?  We should count the failures and include an error count in the result object.

Reply:
Nice catch, something got lost in merge